### PR TITLE
Support TPM seal/unseal on x86/FSP targets

### PR DESCRIPTION
**Pull request by rizlik**

### CODE REVIEW

File: config/examples/kontron_vx3060_s2.config [hunk 1]
```diff
@@ -1,8 +1,8 @@
 ARCH=x86_64
 TARGET=kontron_vx3060_s2
 WOLFBOOT_SMALL_STACK=0
-SIGN=ECC384
-HASH=SHA384
+SIGN=ECC256
```
Comment (dgarske) [inline config/examples/kontron_vx3060_s2.config:4]: We should default to ECC384 for the examples, right?

File: src/boot_x86_fsp.c [hunk 2]
```diff
@@ -340,6 +398,33 @@ static int fsp_silicon_init(struct fsp_info_header *fsp_info, uint8_t *fsp_s_bas
...
+static int self_extend_pcr(void)
+{
+    uint8_t hash[SHA256_DIGEST_SIZE];
+    uint32_t blksz, position = 0;
+    wc_Sha256 sha256_ctx;
+    uint32_t sz;
+    uintptr_t p;
+
+    p  = (uintptr_t)_start_keystore;
+    sz = ((MEMORY_4GB) - (uint64_t)p);
```
Comment (dgarske) [inline src/boot_x86_fsp.c:413]: Is this `MEMORY_4GB` a physical limit for determining end of RAM? Would be nice to document or provide a helper for uses around end of memory location.

File: src/boot_x86_fsp.c [hunk 3]
```diff
@@ -398,6 +494,14 @@ static void memory_ready_entry(void *ptr)
     }
 #endif
 
+#if defined(WOLFBOOT_MEASURED_BOOT)
+    ret = wolfBoot_image_measure((uint8_t*)fsp_s_base - IMAGE_HEADER_SIZE);
+    if (ret != 0) {
+        wolfBoot_printf("Fail to measure FSP_S image\r\n");
```
Comment (dgarske) [inline src/boot_x86_fsp.c:500]: Seeing mixed use of `\r\n` and `ENDLINE`?

File: tools/tpm/policy_sign.c [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    mp_clear(&r);
+    mp_clear(&s);
+    return 0;
+}
+
+#define SHA256_LEN 32
+static int policy_sign(const char *policy_file, const char *key_path) {
+    uint8_t digest[SHA256_LEN];
+    unsigned int digest_sz;
+    int policy_sz = 64;
```
Comment (dgarske) [inline tools/tpm/policy_sign.c:97]: Please cleanup these hard coded 64's. Also at line 122.

File: tools/tpm/policy_sign.c [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    uint8_t buf[1024];
```
Comment (dgarske) [inline tools/tpm/policy_sign.c:98]: Can this be MAX_PATH?

File: tools/tpm/policy_sign.c [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+static int sign_digest(const uint8_t *digest, int digest_sz, uint8_t *signature,
+                       ecc_key *k)
+{
+    int sig_sz = 32;
+    mp_int r, s;
+    WC_RNG rng;
+    int ret;
+
+    ret = wc_InitRng(&rng);
+    if (ret != 0) {
```
Comment (dgarske) [inline tools/tpm/policy_sign.c:75]: Please convert this into a single return and make sure mp_clear is called after mp_init. The sign failure may not call mp_clear which could leak memory for some math implementations.

File: src/x86/ahci.c [hunk 7]
```diff
@@ -185,6 +214,209 @@ void ahci_dump_port(uint32_t base, int i)
...
+                                              const uint8_t *policy,
+                                              int policy_size,
+                                              uint8_t *secret,
+                                              int *secret_size)
+{
+    uint8_t secret_check[WOLFBOOT_MAX_SEAL_SZ];
+    int secret_check_sz;
+    int ret;
+    int r;
+    uint8_t test[32];
```
Comment (dgarske) [inline src/x86/ahci.c:289]: Unused?

File: src/x86/ahci.c [hunk 8]
```diff
@@ -185,6 +214,209 @@ void ahci_dump_port(uint32_t base, int i)
...
+
+    if (ret == 0) {
+        wolfBoot_printf("Secret %d bytes\n", *secret_size);
+        wolfBoot_printf("%s\r\n", secret);
+    }
+
+    return ret;
+}
+
+static int sata_get_unlock_secret(uint8_t *secret, int *secret_size) {
```
Comment (dgarske) [inline src/x86/ahci.c:332]: Function start brace on new line please.

**Review by dgarske — CHANGES_REQUESTED** [review]
Great work on this. A few cleanups.

File: tools/tpm/policy_sign.c [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+            printf("Key file read error!\n");
+            goto exit0;
+        }
+    }
+exit0:
+    fclose(f);
+    return 0;
+}
+
+static int sign_digest(const uint8_t *digest, int digest_sz, uint8_t *signature,
```
[the remark this answers is not in the document]
Comment (rizlik) [reply tools/tpm/policy_sign.c]: Thanks, I remake `tools/tpm/policy_sign.c` based on [https://github.com/wolfSSL/wolfTPM/blob/master/examples/pcr/policy_sign.c](https://github.com/wolfSSL/wolfTPM/blob/master/examples/pcr/policy_sign.c). It still supports ECC256 only, but support for other key types can be easily integrated.

File: src/x86/ahci.c [hunk 10]
```diff
@@ -40,6 +40,35 @@
...
+#if defined(WOLFBOOT_TPM_SEAL)
+#include <image.h>
+#include <tpm.h>
+
+#if defined(WOLFBOOT_FSP)
+#include <stage1.h>
+#endif
+
+#include <wolfssl/wolfcrypt/coding.h>
+/* hardcode to ecc256 algo for now */
```
[the remark this answers is not in the document]
Comment (rizlik) [reply src/x86/ahci.c:52]: I added a `uint32_t` size before the policy in the flash. Who puts the policy in the flash needs to update the size value as well.

File: config/examples/kontron_vx3060_s2.config (hunk 1 above)
Comment (rizlik) [reply config/examples/kontron_vx3060_s2.config:4]: To have ECC384 working right now two signing keys must be added to the keystore, one for authenticating the images and one for authenticating the policy for the sealed secret. I think it's better to postpone an example with different keys from TPM and images in a different PR.

**Review by dgarske — DISMISSED** [review]
Over to @danielinux for final.

File: hal/x86_fsp_qemu_stage1.ld.in [hunk 11]
```diff
@@ -18,22 +19,42 @@ MEMORY
 }
 
 SECTIONS
-{
-    .jmpto32 BOOTLOADER_JUMP32_START :
+{
```
Comment (danielinux) [inline hal/x86_fsp_qemu_stage1.ld.in:22]: remove: whitespaces

File: arch.mk [hunk 12]
```diff
@@ -617,7 +617,7 @@ ifeq ("${FSP}", "1")
               --defsym wb_start_bss=`grep _start_bss ../wolfboot.map | awk '{print $$1}'` \
               --defsym wb_end_bss=`grep _end_bss ../wolfboot.map | awk '{print $$1}'` \
               --defsym _stage2_params=`grep _stage2_params ../wolfboot.map | awk '{print $$1}'`
-    LDFLAGS +=  --no-gc-sections --print-gc-sections -T $(LSCRIPT) -m elf_i386  -Map=loader_stage1.map
+    LDFLAGS +=  --no-gc-sections --print-gc-sections -T $(LSCRIPT) -m elf_i386  -Map=loader_stage1.map -N
```
Comment (danielinux) [inline arch.mk:620]: "-N" : unclear purpose. reconsider?

File: tools/tpm/policy_sign.c [hunk 13]
```diff
@@ -0,0 +1,395 @@
+/* policy_sign.c
+ *
+ * Copyright (C) 2006-2023 wolfSSL Inc.
+ *
+ * This file is part of wolfTPM.
+ *
+ * wolfBoot is free software; you can redistribute it and/or modify
```
Comment (danielinux) [inline tools/tpm/policy_sign.c:7]: mismatch: wolfBoot/wolfTPM in comments. Please uniform.

**Review by danielinux — CHANGES_REQUESTED** [review]
Just a couple of minor fixes

**Review by danielinux — APPROVED** [review]

**Pull request merged**